### PR TITLE
Add support for Azure Container Registry, update Azure dependencies

### PR DESCRIPTION
**Pull request by brendandburns**
@colemickens

### CODE REVIEW

**Comment by k8s-oncall** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37783)

**Comment by brendandburns** [comment]
Work in progress, needs unit tests, and integration with the kubelet binary.

**Comment by colemickens** [comment]
1. I assume that the user has to go to ACR and grant access for the SP that the cluster is using? Presumably it doesn't inherit the RBAC (similar to KV/storage/etc)...

2. Maybe the config parse/environment determination/SPToken creation could be refactored and shared with cloudprovider?

I can spot some compile-time errors, will hold off reviewing until you push more changes.

**Comment by bgrant0607** [comment]
@brendandburns It's really late to be adding this to 1.5. 

cc @saad-ali

**Comment by brendandburns** [comment]
@bgrant0607 it's totally modular and low risk (off everywhere except Azure)

**Comment by bgrant0607** [comment]
@brendandburns This is the exception process for non-bugfix changes:
https://github.com/kubernetes/features/blob/master/EXCEPTIONS.md

**Comment by saad-ali** [comment]
Like Brian said, we are past feature freeze and are only accepting bug fixes for 1.5.0 at the moment. I'm going to remove the 1.5 label for now. This can be considered for a 1.5.x patch release once 1.5.0 is out

**Comment by brendandburns** [comment]
@colemickens test added, code fixed.  Refactor is a todo?

@saad-ali I'll propose a cherry-pick it into the 1.5.1 release.

File: pkg/credentialprovider/azure/azure_credentials.go
Comment (colemickens) [inline pkg/credentialprovider/azure/azure_credentials.go]: Do these wind up exposed on kubelet then?

File: pkg/credentialprovider/azure/azure_credentials.go
Comment (colemickens) [inline pkg/credentialprovider/azure/azure_credentials.go]: Error message looks copy/pasted from above, should probably mention "failed to get service principal oauth token" or something similar.

File: pkg/credentialprovider/azure/azure_credentials.go
Comment (colemickens) [inline pkg/credentialprovider/azure/azure_credentials.go]: related: https://github.com/Azure/go-autorest/issues/102

**Review by colemickens — APPROVED** [review]
few small comments, otherwise looks good

File: pkg/credentialprovider/azure/azure_credentials.go
Comment (brendandburns) [inline pkg/credentialprovider/azure/azure_credentials.go]: Yes (switched to pointers so it will actually work...)

File: pkg/credentialprovider/azure/azure_credentials.go
Comment (brendandburns) [inline pkg/credentialprovider/azure/azure_credentials.go]: oops copy/paste...

File: pkg/credentialprovider/azure/azure_credentials.go
Comment (brendandburns) [inline pkg/credentialprovider/azure/azure_credentials.go]: done.

**Comment by brendandburns** [comment]
Comments addressed, rebased, please take another look.

File: pkg/credentialprovider/azure/azure_credentials.go
Comment (colemickens) [inline pkg/credentialprovider/azure/azure_credentials.go]: Maybe consider a default value? I don't think ACR uses it at all (currently?) and [even Docker moved away from it](https://github.com/docker/docker/issues/20565). I imagine we'd just put a dummy value here most of the time anyway.

**Comment by colemickens** [comment]
One other nit comment.

/lgtm

**Comment by brendandburns** [comment]
@k8s-bot gce etcd3 e2e test this issue: #33380

**Comment by brendandburns** [comment]
re-lgtm on cole's LGTM above.

**Comment by luxas** [comment]
This broke the cross-build, due to that we're vendoring an old version of `github.com/docker/docker/pkg/term/windows/` that was coded against an old version of `github.com/Azure/go-ansiterm`. So now when we bumped the version of `github.com/Azure/go-ansiterm` here, breaking changes in `go-ansiterm` makes `pkg/term/windows` fail so we broke all windows builds at once.

We could of course try to vendor a newer version of `github.com/docker/docker/pkg/term/windows/` alone, but that's risky and might take some time depending on how old version we have vendored now.

I would suggest we revert this ASAP and try merging a new version of this PR again later this week.
In the future, please comment "@k8s-bot build this" on all vendoring changes you make in order to be confident it won't break the queue.

cc @kubernetes/sig-testing @k8s-oncall

**Comment by colemickens** [comment]
@luxas I'm a bit confused. This PR bumps `go-autorest` but not `go-ansiterm`. Or is there a transitive dependency problem?

**Comment by luxas** [comment]
@colemickens This PR does indeed bump `go-ansiterm`: https://github.com/brendandburns/kubernetes/blob/91f19e3dfc1f5c282032122beefa71d7ab026561/Godeps/Godeps.json#L62

That makes the crossbuild fail: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-cross-build/211

The vendored package `github.com/docker/docker/pkg/term/windows` is coded against an old version of `go-ansiterm`, and now when `go-ansiterm` was bumped here and contained breaking changes compared to the earlier version Azure/go-ansiterm@70b2c90b260171e829f1ebd7c17f600c11858dbe, so now `github.com/docker/docker/pkg/term/windows` fails to compile.

Can we revert this temporarily to unblock the crossbuild? We really need the cross-build to be healthy, especially tomorrow when kubeadm is gonna be released.

**Comment by colemickens** [comment]
Sorry, I see it now. I guess if it's blocking a kubeadm release it would make sense to revert to unblock since this is currently targeting `1.5.1` anyway.

Though, I'm wondering, does it build if you revert just the bump to `go-ansiterm`, I'm wondering if @brendandburns bumped all `github.com/Azure/...` packages since `azure-sdk-for-go` and `go-autorest` don't seem to depend on `go-ansiterm`.

**Comment by luxas** [comment]
@colemickens I can test tomorrow, hopefully that will fix the issue. If it doesn't work, I have to revert unfortunately.

Until v1.6, kubeadm releases are dependent on the cross-build for getting the official binary builds, so this is important in that sense to get v1.5 out of the door. (Due to the split-out kubeadm from the monorepo process, etc.)
I'm kind of coordinating the kubeadm release process, see here for more info: https://github.com/kubernetes/kubeadm/issues/70


In the future, we should build some automation that automatically runs the crossbuild ci on presubmit for PRs that touch `vendor/` to avoid this situation. cc @kubernetes/sig-testing

**Comment by luxas** [comment]
Or wait, it might make sense to just revert, otherwise it won't be able to get cherrypicked safely into v1.5.
If this exact PR would go right into the 1.5 branch, it would block the release.
So to minimize the risk, it might be best to do this in a new PR, exactly the same thing but omitting `go-ansiterm` if possible.

WDYT @colemickens about that?

cc @saad-ali

**Comment by colemickens** [comment]
@brendandburns Did you already initiate the cherry pick process for this?

**Comment by luxas** [comment]
If it's gonna be cherrypicked, make sure you include the revert commit in PR as well

**Pull request merged**
